### PR TITLE
Added basic player getters and, operations to FFI

### PR DESCRIPTION
**Pull request by djpiper28**

### CODE REVIEW

**Comment by TylerBloom** [comment]
I plan on putting in lints for enforcing documentation, i.e `#![deny(missing_docs)]`. Since the FFI will be in flex, you might want to turn that off in the FFI files or toggle it if the `ffi` flag is set. For the latter, the following in `lib.rs` should work:
```rust
#![cfg_attr(feature = "ffi", allow(missing_docs))]
```

**Comment by djpiper28** [comment]
(Converted to draft) I am going to implement all basic player operations before any merge.
 - [x] Getting all players for a tournament
 - [x] Adding players to tournaments
 - [ ] Get player with ID
 - [x] Get Player name, game name and, status
 - [ ] Change player game name
 - [ ] Change player name?
 - [ ] Dropping players

File: squire_lib/src/ffi_tournament.rs [hunk 1]
```diff
@@ -53,6 +55,36 @@ impl TournamentId {
         return ptr;
     }
 
+    /// Adds a player to a tournament
+    /// On error a NULL UUID is returned
+    #[no_mangle]
+    pub extern "C" fn tid_add_player(self: Self, __name: *const c_char) -> PlayerId {
```
Comment (TylerBloom) [inline squire_lib/src/ffi_tournament.rs:61]: Is there a reason that these methods take `Self` and not `&Self`?

File: squire_lib/src/ffi_player.rs [hunk 2]
```diff
@@ -1 +1,79 @@
...
+    pub extern "C" fn pid_name(self, tid: TournamentId) -> *const c_char {
+        let player: Player;
+        match self.get_tourn_player(tid) {
+            Some(p) => player = p,
+            None => {
+                return std::ptr::null();
+            }
+        }
+
+        return unsafe { clone_string_to_c_string(player.name) };
```
Comment (TylerBloom) [inline squire_lib/src/ffi_player.rs:40]: Minor nitpick: This return statement isn't required. The same goes for most of these `return`s.

File: squire_lib/src/ffi_tournament.rs [hunk 3]
```diff
@@ -53,41 +55,74 @@ impl TournamentId {
...
+                        return id;
+                    }
+                    Err(t_err) => {
+                        println!("[FFI]: {t_err}");
+                        return PlayerId::new(Uuid::from_bytes(NULL_UUID_BYTES));
+                    }
+                    // We are known the from that the data will take if it exists
+                    // so we can ignore the other outcomes
+                    _ => {
+                        return PlayerId::new(Uuid::from_bytes(NULL_UUID_BYTES));
```
Comment (TylerBloom) [inline squire_lib/src/ffi_tournament.rs:78]: I just double-checked this, and it does work. `Uuid` implements the `Default` trait, which is the same as the null bits. The following is a much more succinct way of getting a null `PlayerId` (or any typed id).
```suggestion
                        return Uuid::default().into();
```

File: squire_lib/src/ffi_player.rs [hunk 4]
```diff
@@ -1 +1,81 @@
...
 
+impl PlayerId {
+    /// Returns the player if it can be found in the tournament
+    fn get_tourn_player(self, tid: TournamentId) -> Option<Player> {
+        let tourn: Tournament;
+        unsafe {
+            match FFI_TOURNAMENT_REGISTRY.get_mut().unwrap().get(&tid) {
+                Some(t) => tourn = t.value().clone(),
+                None => {
+                    println!("Cannot find player in get_tourn_player();");
```
Comment (TylerBloom) [inline squire_lib/src/ffi_player.rs:18]: This print statement is incorrect. The tournament is unable to be found, not the player.

File: squire_lib/src/ffi_player.rs [hunk 5]
```diff
@@ -1 +1,81 @@
...
+
+        return unsafe { clone_string_to_c_string(player.name) };
+    }
+
+    /// Returns the player's game name if they can be found
+    /// NULL is returned on error or, failure to find
+    #[no_mangle]
+    pub extern "C" fn pid_game_name(self, tid: TournamentId) -> *const c_char {
+        let player: Player;
+        match self.get_tourn_player(tid) {
```
Comment (TylerBloom) [inline squire_lib/src/ffi_player.rs:50]: In line with my other comments, these match statements can be merged and/or you can use chained methods on `Option` to boil everything down.

File: squire_lib/src/ffi_player.rs [hunk 6]
```diff
@@ -1 +1,81 @@
...
+            }
+        }
+    }
+
+    /// Returns the player's status if they can be found
+    /// Dropped on error.
+    #[no_mangle]
+    pub extern "C" fn pid_status(self, tid: TournamentId) -> PlayerStatus {
+        let player: Player;
+        match self.get_tourn_player(tid) {
```
Comment (TylerBloom) [inline squire_lib/src/ffi_player.rs:72]: Similar thing. No need for a match statement.
```suggestion
        self.get_tourn_player(tid).map(|p| p.status).unwrap_or(PlayerStatus::Dropped)
```

File: squire_lib/src/ffi_tournament.rs [hunk 7]
```diff
@@ -53,41 +57,75 @@ impl TournamentId {
         return ptr;
     }
 
+    /// Adds a player to a tournament
+    /// On error a NULL UUID is returned
+    #[no_mangle]
+    pub unsafe extern "C" fn tid_add_player(self: Self, __name: *const c_char) -> PlayerId {
```
Comment (TylerBloom) [inline squire_lib/src/ffi_tournament.rs:63]: I don't think this needs to be `unsafe`.

File: squire_lib/src/ffi_tournament.rs [hunk 8]
```diff
@@ -53,41 +57,75 @@ impl TournamentId {
...
+        let name: &str = unsafe { CStr::from_ptr(__name).to_str().unwrap() };
+        let op: TournOp =
+            TournOp::RegisterPlayer(SquireAccount::new(name.to_string(), name.to_string()));
+
+        match FFI_TOURNAMENT_REGISTRY.get_mut().unwrap().get_mut(&self) {
```
Comment (TylerBloom) [inline squire_lib/src/ffi_tournament.rs:68]: You actually don't need this first `get_mut()`. `DashMap` is a concurrent map, so it handles interior mutability for us.
```suggestion
        match FFI_TOURNAMENT_REGISTRY.get().unwrap().get_mut(&self) {
```

File: squire_lib/src/ffi_tournament.rs [hunk 9]
```diff
@@ -53,41 +57,75 @@ impl TournamentId {
...
+            Some(mut t) => {
```
Comment (TylerBloom) [inline squire_lib/src/ffi_tournament.rs:69]: You shouldn't need to have `mut t` here. `mut` is for when the value of a variable changes, but `t` is a `&mut Tournament`. `mut t` implies that you are going to reference a different `Tournament`.
```suggestion
            Some(t) => {
```

File: squire_lib/src/ffi_tournament.rs [hunk 10]
```diff
@@ -53,41 +57,75 @@ impl TournamentId {
...
+                match t.apply_op(op) {
+                    Ok(RegisterPlayer(PlayerIdentifier::Id(id))) => {
+                        return id;
+                    }
+                    Err(t_err) => {
+                        println!("[FFI]: {t_err}");
+                        return PlayerId::new(Uuid::from_bytes(NULL_UUID_BYTES));
```
Comment (TylerBloom) [inline squire_lib/src/ffi_tournament.rs:76]:
```suggestion
                        Uuid::default().into()
```

File: squire_lib/src/ffi_tournament.rs [hunk 11]
```diff
@@ -53,41 +57,75 @@ impl TournamentId {
...
+                    }
+                    // We are known the from that the data will take if it exists
+                    // so we can ignore the other outcomes
+                    _ => {
+                        return Uuid::default().into();
```
Comment (TylerBloom) [inline squire_lib/src/ffi_tournament.rs:81]:
```suggestion
                        Uuid::default().into()
```

File: squire_lib/src/ffi_tournament.rs [hunk 12]
```diff
@@ -53,41 +57,75 @@ impl TournamentId {
...
+                    }
+                };
+            }
+            None => {
+                println!("[FFI]: Cannot find tournament in tid_add_player");
+                return PlayerId::new(Uuid::from_bytes(NULL_UUID_BYTES));
```
Comment (TylerBloom) [inline squire_lib/src/ffi_tournament.rs:87]:
```suggestion
                        Uuid::default().into()
```

File: squire_lib/src/ffi_tournament.rs [hunk 13]
```diff
@@ -53,41 +57,75 @@ impl TournamentId {
...
+            }
+        }
+    }
+
     /// Returns the name of a tournament
     /// Returns NULL if an error happens
     /// This is heap allocated, please free it
     #[no_mangle]
-    pub extern "C" fn tid_name(self: Self) -> *const c_char {
+    pub unsafe extern "C" fn tid_name(self: Self) -> *const c_char {
```
Comment (TylerBloom) [inline squire_lib/src/ffi_tournament.rs:96]:
```suggestion
    pub extern "C" fn tid_name(self: Self) -> *const c_char {
```

File: squire_lib/src/ffi_tournament.rs [hunk 14]
```diff
@@ -53,41 +57,75 @@ impl TournamentId {
...
         let tourn: Tournament;
-        match FFI_TOURNAMENT_REGISTRY.get().unwrap().get(&self) {
+        match FFI_TOURNAMENT_REGISTRY.get_mut().unwrap().get(&self) {
             Some(t) => tourn = t.value().clone(),
```
Comment (TylerBloom) [inline squire_lib/src/ffi_tournament.rs:99]: This clone is not needed. Like my comments on the player ffi, you can remove this match statement and/or condense this function to only be the `match` and have no `return`.

File: squire_lib/src/ffi_tournament.rs [hunk 15]
```diff
@@ -53,41 +57,75 @@ impl TournamentId {
...
-        return unsafe { clone_string_to_c_string(tourn.name) };
+        return clone_string_to_c_string(tourn.name);
     }
 
     /// Returns the format of a tournament
     /// Returns NULL if an error happens
     /// This is heap allocated, please free it
     #[no_mangle]
-    pub extern "C" fn tid_format(self: Self) -> *const c_char {
+    pub unsafe extern "C" fn tid_format(self: Self) -> *const c_char {
```
Comment (TylerBloom) [inline squire_lib/src/ffi_tournament.rs:112]: Same comments as before.
```suggestion
    pub extern "C" fn tid_format(self: Self) -> *const c_char {
```

File: squire_lib/src/ffi_tournament.rs [hunk 16]
```diff
@@ -215,8 +255,16 @@ impl TournamentId {
 
     /// Closes a tournament removing it from the internal FFI state
     #[no_mangle]
-    pub extern "C" fn close_tourn(self: Self) {
-        FFI_TOURNAMENT_REGISTRY.get().unwrap().remove(&self);
+    pub unsafe extern "C" fn close_tourn(self: Self) -> bool {
+        match FFI_TOURNAMENT_REGISTRY.get_mut().unwrap().remove(&self) {
```
Comment (TylerBloom) [inline squire_lib/src/ffi_tournament.rs:259]: You probably want to close the tournament with `TournOp::Close()` and remove the file.

File: squire_lib/src/ffi_tournament.rs [hunk 17]
```diff
@@ -280,21 +337,25 @@ pub extern "C" fn load_tournament_from_file(__file: *const c_char) -> Tournament
...
 
-    // Cannot open the same tournament twice
-    if FFI_TOURNAMENT_REGISTRY
-        .get()
-        .unwrap()
-        .contains_key(&tournament.id)
-    {
-        println!("Input file is already open");
-        return TournamentId::new(Uuid::from_bytes(NULL_UUID_BYTES));
+    unsafe {
```
Comment (TylerBloom) [inline squire_lib/src/ffi_tournament.rs:340]: Unnecessary `unsafe`

File: squire_lib/src/ffi_tournament.rs [hunk 18]
```diff
@@ -280,21 +337,25 @@ pub extern "C" fn load_tournament_from_file(__file: *const c_char) -> Tournament
...
     let tid: TournamentId = tournament.id.clone();
-    FFI_TOURNAMENT_REGISTRY
-        .get()
-        .unwrap()
-        .insert(tid, tournament.clone());
+    unsafe {
+        FFI_TOURNAMENT_REGISTRY
+            .get_mut()
+            .unwrap()
+            .insert(tid, tournament.clone());
```
Comment (TylerBloom) [inline squire_lib/src/ffi_tournament.rs:357]: Unnecessary `clone`. you can just return `tid`. This is also in an unnecessary `unsafe` block.

File: squire_lib/src/ffi_tournament.rs [hunk 19]
```diff
@@ -331,13 +392,17 @@ pub extern "C" fn new_tournament_from_settings(
...
+        judges: HashMap::new(),
+        admins: HashMap::new(),
     };
     let tid: TournamentId = tournament.id;
 
-    FFI_TOURNAMENT_REGISTRY
-        .get()
-        .unwrap()
-        .insert(tid, tournament.clone());
+    unsafe {
```
Comment (TylerBloom) [inline squire_lib/src/ffi_tournament.rs:400]: Same comment as above.

**Review by TylerBloom — COMMENTED** [review]
It's looking good overall. There are just a lot of missed Rust-isms. I didn't leave a comment on all of them because I assume you'll be able to see the pattern. These are all definitely worth changing as it makes the code more readable and will make your first pass of future methods have less boilerplate.

I would recommend running `cargo clippy --features ffi`. Clippy is a great linting tool that will catch a lot of these.

File: squire_lib/src/ffi_player.rs [hunk 20]
```diff
@@ -1 +1,79 @@
...
+            None => {
+                return std::ptr::null();
+            }
+        }
+    }
+
+    /// Returns the player's status if they can be found
+    /// Dropped on error.
+    #[no_mangle]
+    pub extern "C" fn pid_status(self, tid: TournamentId) -> PlayerStatus {
```
[the remark this answers is not in the document]
Comment (djpiper28) [reply squire_lib/src/ffi_player.rs:70]: The C side will throw uncaught case in switch statements, meaning that it is easy to spot all missing occurances and, update it.

File: squire_lib/src/ffi_tournament.rs (hunk 1 above)
Comment (djpiper28) [reply squire_lib/src/ffi_tournament.rs]: rust didn't complain about no & is the reason

Comment (TylerBloom) [reply squire_lib/src/ffi_tournament.rs]: Ah, gotcha. That makes sense I guess. `Uuid` (and by extension `TypeId`) implements `Copy`. You should be able to get rid of `: Self` and have the method take `self` (or `&self`).

File: squire_lib/src/ffi.rs [hunk 21]
```diff
@@ -13,15 +13,19 @@ pub const NULL_UUID_BYTES: [u8; 16] = [0; 16];
...
 #[no_mangle]
-pub extern "C" fn init_squire_ffi() {
+pub unsafe extern "C" fn init_squire_ffi() {
     let map: DashMap<TournamentId, Tournament> = DashMap::new();
     FFI_TOURNAMENT_REGISTRY.set(map).unwrap();
 }
 
 /// Helper function for cloning strings
+/// Returns NULL on error
 pub unsafe fn clone_string_to_c_string(s: String) -> *mut c_char {
```
[the remark this answers is not in the document]
Comment (djpiper28) [reply squire_lib/src/ffi.rs]: Resolved in 858b58120b37624518869cf661de1daf9623d45d

File: squire_lib/src/ffi_player.rs (hunk 2 above)
Comment (djpiper28) [reply squire_lib/src/ffi_player.rs:42]: Tried it, it didn't work for some reason.

File: squire_lib/src/ffi_player.rs [hunk 22]
```diff
@@ -1 +1,81 @@
...
+            match FFI_TOURNAMENT_REGISTRY.get_mut().unwrap().get(&tid) {
+                Some(t) => tourn = t.value().clone(),
+                None => {
+                    println!("Cannot find player in get_tourn_player();");
+                    return None;
+                }
+            }
+        }
+
+        match tourn.player_reg.get_player(&PlayerIdentifier::Id(self)) {
```
[the remark this answers is not in the document]
Comment (djpiper28) [reply squire_lib/src/ffi_player.rs]: Resolved in 1ee32e5ead528cd5cc6e9fe4fb0310ea0012c917

File: squire_lib/src/ffi_player.rs [hunk 23]
```diff
@@ -1 +1,81 @@
...
+        }
+    }
+
+    /// Returns the player name if they can be found
+    /// NULL is returned on error or, failure to find
+    #[no_mangle]
+    pub extern "C" fn pid_name(self, tid: TournamentId) -> *const c_char {
+        let player: Player;
+        match self.get_tourn_player(tid) {
+            Some(p) => player = p,
```
[the remark this answers is not in the document]
Comment (djpiper28) [reply squire_lib/src/ffi_player.rs]: Resolved in 14971e77671a61e5833da56a61a6ad4ed4021e3b

File: squire_lib/src/ffi_player.rs (hunk 6 above)
Comment (djpiper28) [reply squire_lib/src/ffi_player.rs:65]: Resolved in 14971e77671a61e5833da56a61a6ad4ed4021e3b

File: squire_lib/src/ffi_tournament.rs (hunk 18 above)
Comment (djpiper28) [reply squire_lib/src/ffi_tournament.rs]: Resolved in acaed8005beb87ce9eb8914db6a871079f558b76

File: squire_lib/src/ffi_player.rs (hunk 4 above)
Comment (djpiper28) [reply squire_lib/src/ffi_player.rs]: 00f17b7c64bb77bbeda4f060c47348b9c32df51f

File: squire_lib/src/ffi_tournament.rs (hunk 9 above)
Comment (djpiper28) [reply squire_lib/src/ffi_tournament.rs:71]: error[E0596]: cannot borrow `t` as mutable, as it is not declared as mutable
  --> squire_lib/src/ffi_tournament.rs:72:23
   |
71 |             Some(t) => {
   |                  - help: consider changing this to be mutable: `mut t`
72 |                 match t.apply_op(op) {
   |                       ^^^^^^^^^^^^^^ cannot borrow as mutable

For more information about this error, try `rustc --explain E0596`.
error: could not compile `squire_lib` due to previous error

File: squire_lib/src/ffi_tournament.rs (hunk 10 above)
Comment (djpiper28) [reply squire_lib/src/ffi_tournament.rs]: cd045b78485f0cea4edd9e28e5d188aa90d3a122

File: squire_lib/src/ffi_tournament.rs (hunk 11 above)
Comment (djpiper28) [reply squire_lib/src/ffi_tournament.rs:83]: cd045b78485f0cea4edd9e28e5d188aa90d3a122

File: squire_lib/src/ffi_tournament.rs (hunk 12 above)
Comment (djpiper28) [reply squire_lib/src/ffi_tournament.rs]: cd045b78485f0cea4edd9e28e5d188aa90d3a122

File: squire_lib/src/ffi_tournament.rs (hunk 9 above)
Comment (TylerBloom) [reply squire_lib/src/ffi_tournament.rs:71]: Oh, right! `DashMap::get_mut` doesn't return a reference. It returns a guard that gives out a reference.

**Comment by TylerBloom** [comment]
This has been manually rebased with main.

**Pull request closed without merging**